File: charge.cpp
#include <bits/stdc++.h>
using namespace std;

const double INF = 1e9;
const double time_per_distance = 0.1;
const double distance_per_charge = 3;
const double time_per_charge = 0.05;
const int charge_quantums = 8;
const double weight_penalty_factor = 0.0001;
const double base_consumption_per_dist = 1.0 / distance_per_charge;

// Helper Calculations
static double time_for_distance(double dis) {
    return time_per_distance * dis;
}

static int charge_loss_for_distance(double dis, double cargo_weight) {
    double effective_consumption = base_consumption_per_dist + (cargo_weight * weight_penalty_factor);
    // Charge used is the metric we want to minimize
    double charge_loss = dis * effective_consumption;
    return ceil(charge_loss);
}

// The State for Dijkstra
struct State {
    int u;             
    int bat;
    double time;
    int total_charge;

    bool operator>(const State& other) const {
        return total_charge > other.total_charge;
    }
};

void path_checker_charge(
    int delivery_points,
    int charging_stations,
    vector<pair<double, double>> &times, // Time Windows
    vector<double> &distance,            // Dist A -> B
    vector<vector<double>> &dis_c,       // Dist A -> Station -> B logic needs care
    vector<double> &cargo_weight
) {
    // 1. DP Table for Pruning: Stores MIN TIME for a given state
    // We minimize charge via PQ, but we prune if we are too slow.
    vector<vector<double>> best_time(delivery_points, vector<double>(charge_quantums + 1, INF));

    // 2. Dijkstra Priority Queue
    priority_queue<State, vector<State>, greater<State>> pq;

    // Initial State: Node 0, Max Charge, Time 0, Total Consumed 0
    pq.push({0, charge_quantums, 0.0, 0});
    best_time[0][charge_quantums] = 0.0;

    double min_total_charge = INF, time_taken_for_min_charge = INF;

    while (!pq.empty()) {
        State current = pq.top();
        pq.pop();

        int u = current.u;
        int bat = current.bat;
        double time = current.time;
        int cost = current.total_charge;

        // --- GOAL CHECK ---
        if (u == delivery_points - 1) {
            min_total_charge = cost;
            time_taken_for_min_charge = time;
            break;
        }

        // --- PRUNING ---
        if (time > best_time[u][bat]) continue;

        // Current payload weight
        double cur_w = cargo_weight[u];
        
        // --- OPTION 1: Drive DIRECTLY to Next Node (u -> u+1) ---
        if (u < delivery_points - 1) {
            double dist_next = distance[u];
            int loss = charge_loss_for_distance(dist_next, cur_w);
            
            if (bat >= loss) {
                int next_bat = bat - loss;
                double travel_time = time_for_distance(dist_next);
                double arrival_time = time + travel_time;
                
                // Check Time Window for Next Node
                double window_start = times[u+1].first;
                double window_end = times[u+1].second;

                double final_arrival = max(arrival_time, window_start);

                if (final_arrival <= window_end) {
                    if (final_arrival < best_time[u+1][next_bat]) {
                        best_time[u+1][next_bat] = final_arrival;
                        pq.push({u + 1, next_bat, final_arrival, cost + loss});
                    }
                }
            }
        }

        // --- OPTION 2: Visit Charging Station (u -> Station -> u+1) ---
        if (u < delivery_points - 1) {
            for (int s = 0; s < charging_stations; s++) {
                // Dist: Node -> Station
                double d1 = dis_c[u][s]; 
                double d2 = dis_c[u+1][s]; 

                int loss1 = charge_loss_for_distance(d1, cur_w);
                int loss2 = charge_loss_for_distance(d2, cur_w);

                if (bat >= loss1) {
                    int bat_at_station = bat - loss1;
                    
                    for (int target_bat = bat_at_station; target_bat <= charge_quantums; target_bat++) {
                        int added_charge = target_bat - bat_at_station;
                        
                        // Can we leave station and reach next node?
                        if (target_bat >= loss2) {
                            int final_bat = target_bat - loss2;
                            
                            double t_travel = time_for_distance(d1) + time_for_distance(d2);
                            double t_charge = added_charge * time_per_charge;
                            double arrival_time = time + t_travel + t_charge;

                            double window_start = times[u+1].first;
                            double window_end = times[u+1].second;
                            double final_arrival = max(arrival_time, window_start);

                            if (final_arrival <= window_end) {
                                if (final_arrival < best_time[u+1][final_bat]) {
                                    best_time[u+1][final_bat] = final_arrival;
                                    pq.push({u + 1, final_bat, final_arrival, cost + loss1 + loss2});
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    if (min_total_charge == INF) cout << "IMPOSSIBLE\n";
    else{
        cout << "Minimum Total Charge Used: " << min_total_charge << "\n";
        cout << "Time Taken For Minimum Total Charge: " << time_taken_for_min_charge << "\n";
    }
}

void path_checker_charge_external(
    int delivery_points,
    int charging_stations,
    vector<pair<double, double>> times,
    vector<double> distance,
    vector<vector<double>> dis_c,
    vector<double> delivery_weights
){
    cout << times.back().first << endl;
    times.insert(times.begin(), {0, INF});
    times.push_back({0, INF});
    delivery_points += 2;

    // add one more dis_c row for final depot return (reuse first row)
    dis_c.push_back(dis_c[0]);

    vector <double> cargo_weight(delivery_points - 1);

    double total_weight = 0;
    for (double w : delivery_weights) {
        total_weight += w;
    }

    cargo_weight[0] = total_weight;
    for (int i = 1; i < delivery_points - 1; i++) {
        cargo_weight[i] = cargo_weight[i-1] - delivery_weights[i-1];
    }

    path_checker_charge(delivery_points, charging_stations, times, distance, dis_c, cargo_weight);
}

int main() {
    int delivery_points, charging_stations;
    cin >> delivery_points >> charging_stations;

    // assuming the time array provided is the delivery sequence the EV guy will be following
    // since it also contains the return node the time for final will be (0,1e9)
    vector<pair<double,double>> times(delivery_points);
    for (int i = 0; i < delivery_points; i++) cin >> times[i].first >> times[i].second;

    vector <double> delivery_weights(delivery_points);
    for (int i = 0; i < delivery_points; i++) cin >> delivery_weights[i];

    // take in two extra distances, depot to first city and last city to depot
    vector<double> distance(delivery_points + 1);
    for (int i = 0; i <= delivery_points; i++) cin >> distance[i];

    // distance of charging stations from node, here first array will be depot to charging station, so one extra
    vector<vector<double>> dis_c(delivery_points + 1, vector<double>(charging_stations));
    for (int i = 0; i <= delivery_points; i++)
        for (int j = 0; j < charging_stations; j++)
            cin >> dis_c[i][j];

    // add depot start + end
    times.insert(times.begin(), {0, INF});
    times.push_back({0, INF});
    delivery_points += 2;

    // add one more dis_c row for final depot return (reuse first row)
    dis_c.push_back(dis_c[0]);

    vector <double> cargo_weight(delivery_points - 1);

    double total_weight = 0;
    for (double w : delivery_weights) {
        total_weight += w;
    }

    cargo_weight[0] = total_weight;
    for (int i = 1; i < delivery_points - 1; i++) {
        cargo_weight[i] = cargo_weight[i-1] - delivery_weights[i-1];
    }

    path_checker_charge(delivery_points, charging_stations, times, distance, dis_c, cargo_weight);
}

File: charge.py
import math
import heapq
import numpy as np
from dataclasses import dataclass

INF = 1e9
time_per_distance = 0.1
distance_per_charge = 5
time_per_charge = 0.05
charge_quantums = 500
MAX_DIS = charge_quantums * distance_per_charge
base_consumption_per_dist = 1 / distance_per_charge
weight_penalty_factor = 0.01

@dataclass(order=False)
class State:
    node: int
    current_charge: int
    time_taken: float
    total_charge_consumed: int

    def __lt__(self, other):
        if not isinstance(other, State):
            return NotImplemented
        if self.total_charge_consumed != other.total_charge_consumed:
            return self.total_charge_consumed < other.total_charge_consumed
        else:
            return self.time_taken < other.time_taken

def _time_for_distance(dist: float):
    return time_per_distance * dist

def _charge_loss_for_distance(dist: float, weight: float) -> int:
    consumption_factor = base_consumption_per_dist + weight * weight_penalty_factor
    return math.ceil(dist * consumption_factor)

def _min_charge(
    delivery_points: int,
    charging_stations: int,
    time_windows: list,
    distance_between_points: list,
    distance_to_stations: list,
    cumulative_weight: list,
):
    min_time = np.full((delivery_points, charge_quantums + 1), INF)
    pq = []
    heapq.heappush(pq, State(0, charge_quantums, 0.0, 0))
    min_time[0][charge_quantums] = 0

    min_total_charge = INF 
    total_time_taken_for_min_charge = INF

    while len(pq):
        current_state: State = heapq.heappop(pq)
        # print(current_state)

        if current_state.node == (delivery_points - 1):
            min_total_charge = current_state.total_charge_consumed
            total_time_taken_for_min_charge = current_state.time_taken
            break

        if current_state.time_taken > min_time[current_state.node][current_state.current_charge]:
            continue

        if current_state.node < (delivery_points - 1):

            # A -> B
            charge_loss_to_next = _charge_loss_for_distance(
                distance_between_points[current_state.node],
                cumulative_weight[current_state.node]
            )

            if current_state.current_charge >= charge_loss_to_next:
                travel_time = _time_for_distance(distance_between_points[current_state.node])
                arrival_time = max(time_windows[current_state.node+1][0], current_state.time_taken + travel_time)
                next_charge = current_state.current_charge - charge_loss_to_next

                if arrival_time <= time_windows[current_state.node+1][1]:
                    if arrival_time < min_time[current_state.node+1][next_charge]:
                        min_time[current_state.node+1][next_charge] = arrival_time
                        heapq.heappush(pq, State(
                            current_state.node + 1,
                            next_charge,
                            arrival_time,
                            current_state.total_charge_consumed + charge_loss_to_next
                        ))
            
            # A->S->B
            for station in range(charging_stations):
                charge_loss_to_station = _charge_loss_for_distance(
                    distance_to_stations[current_state.node][station],
                    cumulative_weight[current_state.node]
                )
                charge_loss_from_station_to_next_node = _charge_loss_for_distance(
                    distance_to_stations[current_state.node + 1][station],
                    cumulative_weight[current_state.node]
                )

                if current_state.current_charge >= charge_loss_to_station:
                    charge_at_station = current_state.current_charge - charge_loss_to_station
                    time_to_station = _time_for_distance(distance_to_stations[current_state.node][station])
                    time_from_station_to_next_node = _time_for_distance(distance_to_stations[current_state.node + 1][station])
                    for charge_added in range(0, charge_quantums - charge_at_station + 1):
                        time_to_charge = charge_added * time_per_charge
                        arrival_time = max(
                            time_windows[current_state.node+1][0],
                            current_state.time_taken + time_to_station + time_to_charge + time_from_station_to_next_node
                        )
                        if arrival_time <= time_windows[current_state.node+1][1]:
                            next_charge = charge_at_station + charge_added - charge_loss_from_station_to_next_node
                            if next_charge < 0:
                                continue
                            if arrival_time < min_time[current_state.node+1][next_charge]:
                                min_time[current_state.node+1][next_charge] = arrival_time
                                heapq.heappush(pq, State(
                                    current_state.node + 1,
                                    next_charge,
                                    arrival_time,
                                    current_state.total_charge_consumed + charge_loss_to_station + charge_loss_from_station_to_next_node
                                ))
    
    if min_total_charge == INF:
        print("IMPOSSIBLE")
    else:
        print("Minimum total charge used: ", min_total_charge)
        print("Time taken for minimum charge route: ", total_time_taken_for_min_charge)
    
    return min_total_charge

def minimum_charge(
    delivery_points: int,
    charging_stations: int,
    time_windows: list,
    distance_between_points: list,
    distance_to_stations: list,
    cargo_weight: list,
):
    # preprocess
    delivery_points += 2
    # print("time: ", len(time_windows))
    # print(time_windows)
    time_windows.insert(0, [0, INF])
    time_windows.append([0, INF])
    distance_to_stations.append(distance_to_stations[0])

    total_weight = sum(cargo_weight)

    cumulative_weights = []
    cumulative_weights.append(total_weight)

    for weight in cargo_weight:
        total_weight -= weight
        cumulative_weights.append(total_weight)

    # print("weight: ", len(cumulative_weights))
    # print("node: ", delivery_points)
    # print("time: ", len(time_windows))

    return _min_charge(
        delivery_points,
        charging_stations,
        time_windows,
        distance_between_points,
        distance_to_stations,
        cumulative_weights
    )
    